java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -696,21 +716,21 @@ private <T extends IncrementalTask> void createTasks(Datastore datastore,
       // We should have way to inject the "shard-id" to the task.
       RetryExecutor.call(FOREVER_RETRYER, () -> {
         IncrementalTaskId taskId = IncrementalTaskId.of(jobId, finalTaskNumber);
-        Transaction tx = datastore.newTransaction();
+        TxnWrapper txnWrapper = TxnWrapper.of(datastore.newTransaction(), taskQueue);
         try {
-          IncrementalTaskState<T> taskState = lookupTaskState(tx, taskId);
+          IncrementalTaskState<T> taskState = lookupTaskState(txnWrapper, taskId);
           if (taskState != null) {
             log.info(jobId + ": Task already exists: " + taskState);
             return null;
           }
           taskState = IncrementalTaskState.create(taskId, jobId, startTime, initialTask);
           ShardRetryState<T> retryState = ShardRetryState.createFor(taskState);
-          tx.put(taskState.toEntity(tx),
-            retryState.toEntity(tx));
-          tx.commit();
-          scheduleWorkerTask(settings, taskState, null);
+          Transaction tx = txnWrapper.getDsTransaction();
+          tx.put(taskState.toEntity(tx), retryState.toEntity(tx));
```

Comment:
`taskState`/`retryState` probably both implement `PipelineObject`, so syntactic sugar could have direct `put(PipelineObject... )` in `TxnWrapper` interface that calls back to underlyin `toEntity(Datastore.Transaction)` methods. in that sense, may be more than just syntactic sugar - but actually better encapsulation.

Reply:
just implemented Transaction and delegate calls, so simpler